jaeger-zipkin/src/main/java/io/jaegertracing/zipkin/V2SpanConverter.java

```diff
@@ -0,0 +1,161 @@
+            .name(span.getOperationName())
+            .parentId(Long.toHexString(context.getParentId()))
+            .debug(context.isDebug())
+            .localEndpoint(host)
+            .remoteEndpoint(peerEndpoint)
+            .kind(convertKind(span.getTags().get(Tags.SPAN_KIND.getKey())))
+            .timestamp(span.getStart())
+            .duration(span.getDuration());
+
+    buildAnnotations(span, builder);
+    buildTags(span, builder);
+
+    return builder.build();
+  }
+
+  private static zipkin2.Span.Kind convertKind(Object kind) {
+    if (Tags.SPAN_KIND_SERVER.equals(kind)) {
+      return zipkin2.Span.Kind.SERVER;
+    } else if (Tags.SPAN_KIND_CLIENT.equals(kind)) {
+      return zipkin2.Span.Kind.CLIENT;
+    } else if (Tags.SPAN_KIND_CONSUMER.equals(kind)) {
+      return zipkin2.Span.Kind.CONSUMER;
+    } else if (Tags.SPAN_KIND_PRODUCER.equals(kind)) {
+      return zipkin2.Span.Kind.PRODUCER;
+    } else {
+      return null;
+    }
+  }
+
+  private static void buildAnnotations(Span span, zipkin2.Span.Builder builder) {
+    List<LogData> logs = span.getLogs();
+    if (logs != null) {
+      for (LogData logData : logs) {
+        String logMessage = logData.getMessage();
+        Map<String, ?> logFields = logData.getFields();
+        if (logMessage != null) {
+          builder.addAnnotation(logData.getTime(), logMessage);
+        } else if (logFields != null) {
+          builder.addAnnotation(logData.getTime(), gson.toJson(logFields));
+        }
+      }
+    }
+  }
+
+  private static void buildTags(Span span, zipkin2.Span.Builder builder) {
+    Map<String, Object> tags = span.getTags();
+    boolean firstSpanInProcess = span.getReferences().isEmpty() || ConverterUtil.isRpcServer(span);
+
+    if (firstSpanInProcess) {
+      Map<String, ?> processTags = span.getTracer().tags();
+      // add tracer tags to first zipkin span in a process but remove "ip" tag as it is
+      // taken care of separately.
+      for (Map.Entry<String, ?> entry : processTags.entrySet()) {
+        String tagKey = entry.getKey();
+        if (!Constants.TRACER_IP_TAG_KEY.equals(tagKey)) {
+          Object tagValue = entry.getValue();
+          // add a tracer. prefix to process tags for zipkin
+          builder.putTag("tracer." + tagKey, tagValue.toString());
+        }
+      }
+    }
+
+    if (tags != null) {
+      for (Map.Entry<String, Object> entry : tags.entrySet()) {
+        String tagKey = entry.getKey();
+        // Every value is converted to string because zipkin search doesn't
+        // work well with ints, and bytes.
+        Object tagValue = entry.getValue();
+        builder.putTag(tagKey, tagValue.toString());
+      }
+    }
+  }
+
+  private static InetAddress convertIp(int ip) {
+    byte[] bytes = ByteBuffer.allocate(4).putInt(ip).array();
+    try {
+      return InetAddress.getByAddress(bytes);
+    } catch (UnknownHostException e) {
+      return null;
```

Comment:
IIRC, this *should* never happen but would be worth adding a log statement here.